related #4274 - added option to verify server certificate with a specific CA

olia-dev opened the pull request:
##### SUMMARY

Per #4274 - added the option to upload a CA certificate to verify connections to the vault server. Code is mostly similar to the one for conjur.py.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - UI

##### AWX VERSION
```
awx: 6.0.0
```

Commit:
related #4274 - added option to verify server certificate with a specific CA

jakemcdermott commented on awx/main/credential_plugins/hashivault.py:
```diff
@@ -93,16 +120,27 @@ def kv_backend(**kwargs):
+                '/'.join([url, mount_point, 'data'] + path).rstrip('/'),
+                params=params,
+                timeout=30
+            )
```
I'm concerned about the duplication of the request building logic for these, _particularly_ the url building, timeout, and other common params. We've had some tricky bug fixes involving these params in the past and so I think the safe thing to do here is to keep them as de-duplicated as possible.

Let's _build up to_ the request args and keyword args like we do in the other plugins:

```python
request_url = '/'.join([url, mount_point, 'data'] + path).rstrip('/')
request_kwargs = {'params': params, 'timeout': 30 }

if cacert:
    request_kwargs['verify'] = create_temporary_fifo(cacert.encode())

response = sess.get(request_url, **request_kwargs)

```

Commit:
related #4274 - added option to verify server certificate with a specific CA

Commit:
related #4274 - added option to verify server certificate with a specific CA (fixed errors)

olia-dev commented:
Hi @jakemcdermott 
thanks for looking over my pull request and your comments. I think your points are correct, and I've modified the code accordingly - could you please check it again?

Another question, while working on integration HashiCorp Vault in our tower setup (and after fixing the SSL problem), I've stumbled upon a second problem and I don't know if I should open a second PR for that - or if I can just add it to this one, since the change is in the same area of the AWX code?

Anyway, what I found (I'll copy that to a new PR if I need to open one):

The URL for Vault K/V (api version: v2) needs '/data/' in it's URL, which gets added on Line 125 but not always at the correct position:
```
    124         # https://www.vaultproject.io/api/secret/kv/kv-v2.html#read-secret-version
    125         request_url = '/'.join([url, mount_point, 'data'] + path).rstrip('/')
    126         response = sess.get(request_url, **request_kwargs)
```
It only gets inserted in the correct place if your (complete) secret path is "mount-name/secret-name", but in our case it is "team-name/application/secret-name".
So the mount_point is actually "team-name/application", and the logic of hashivault.py inserts the data in between those two.

My change would be, for Line 120, old:
```
120             mount_point, *path = pathlib.Path(secret_path.lstrip(os.sep)).parts
```
New:
```
120             *mount_point, path = pathlib.Path(secret_path.lstrip(os.sep)).parts
```
Thus '/data/' would always be inserted before the last slash.

Few more thoughts:
1) Add a check for /data/ in the secret_path, and only add if it is not present
2) Write a validator for secret_path, so that it requires at least "value1/value2"
3) Is the try/exception logic actually needed, especially if 2) would be implemented? I did not hit it in all my tests, but I'm not that experienced of a python programmer?

ryanpetrello commented on awx/main/credential_plugins/hashivault.py:
```diff
@@ -73,36 +82,55 @@
 hashi_ssh_inputs['required'].extend(['public_key', 'role'])
 
 
+def create_temporary_fifo(data):
```
It looks like we've got this function duplicated several places now (conjur, aim, and now here), and it _looks_ identical to me.  Could we place this function in:

`awx/main/utils/common.py`

...and import it here (and in the other plugins, too)?

olia-dev replied:
Should be identical, at least the hashivault.py to the one in conjur.py - since I just copy&pasted it ;-)

jakemcdermott commented:
@olia-dev  
The `/data` path bug you're referring to sounds an awful lot like something we might have fixed recently. https://github.com/ansible/awx/pull/4175

Does your awx deployment have the code changes from that pr? (we merged it in about a week ago)

ryanpetrello commented:
Agreed with @jakemcdermott, sounds like you might want to rebase `devel`.

Commit:
related #4274 - moved function 'create_temporary_fifo' to 'awx/main/utils/common.py' and referenced it in other plugins

Signed-off-by: olia-dev <olia-dev@ktah.net>

olia-dev commented:
@ryanpetrello 
Moved the function `create_temporary_fifo` to 
```
awx/main/utils/common.py
```
And modified the other plugins to reference it.

@jakemcdermott 
I did not know about #4175 specifically, but I've just checked and yes those changes are already included in our installation (and in the branch for this PR).

I'm basically running into the error you referenced in your comment here: https://github.com/ansible/awx/pull/4175#issuecomment-507307284

I've written a fix for that (and use it in our tower installation), but so far haven't included it in this branch. The diff would be:
```
--- hashivault.py	2019-07-10 11:08:50.241034255 +0200
+++ hashivault.py.patched	2019-07-10 11:46:58.644149873 +0200
@@ -104,12 +104,12 @@
         if kwargs.get('secret_version'):
             request_kwargs['params'] = {'version': kwargs['secret_version']}
         try:
-            mount_point, *path = pathlib.Path(secret_path.lstrip(os.sep)).parts
+            *mount_point, path = pathlib.Path(secret_path.lstrip(os.sep)).parts
             '/'.join(path) 
         except Exception:
             mount_point, path = secret_path, []
         # https://www.vaultproject.io/api/secret/kv/kv-v2.html#read-secret-version
-        request_url = '/'.join([url, mount_point, 'data'] + path).rstrip('/')
+        request_url = '/'.join([url] + mount_point + ['data', path]).rstrip('/')
         response = sess.get(request_url, **request_kwargs)
 
         response.raise_for_status()
```

Should I add this to this PR, or open a separate one?

Commit:
related #4274 - moved function 'create_temporary_fifo' to 'awx/main/utils/common.py' and referenced it in other plugins (fixed errors)

Commit:
related #4274 - moved function 'create_temporary_fifo' to 'awx/main/utils/common.py' and referenced it in other plugins (fixed errors)

jakemcdermott commented:
@olia-dev 
Excellent! Thanks for identifying a fix for https://github.com/ansible/awx/pull/4175#issuecomment-507307284. Let's include that in a separate PR so we can focus our review and testing.

jakemcdermott commented:
recheck

jakemcdermott commented:
@olia-dev I think this looks good - Over the next day or so, there's a few checks I'll need to run with some of our tooling, once those are done I'll approve and merge. Thanks for the contribution.

jakemcdermott reviewed (approved):
looks good to me - thanks!
